astroquery/cadc/cadctap/core.py

```diff
@@ -0,0 +1,409 @@
+        when saving if you don't set the format
+        """
+        if jobid is None:
+            if verbose:
+                print("No job identifier found")
+            return None
+        subContext = "async/" + str(jobid)
+        response = self._Tap__connHandler.execute_get(subContext)
+        if verbose:
+            print(response.status, response.reason)
+            print(response.getheaders())
+        isError = self._Tap__connHandler.check_launch_response_status(response,
+                                                                      verbose,
+                                                                      200)
+        if isError:
+            if verbose:
+                print(response.reason)
+            raise requests.exceptions.HTTPError(response.reason)
+            return None
+        # parse job
+        jsp = JobSaxParserCadc(async_job=True)
+        job = jsp.parseData(response)[0]
+        job.connHandler = self._Tap__connHandler
+        suitableOutputFile = self._Tap__getSuitableOutputFile(
+            job.async_,
+            None,
+            response.getheaders(),
+            isError,
+            job.parameters['FORMAT'])
+        job.outputFile = suitableOutputFile
+        job.parameters['format'] = job.parameters['FORMAT']
+        # load results
+        job.get_results(verbose)
+        return job
+
+    def save_results(self, job, filename, verbose=False):
+        """Saves job results
+        Parameters
+        ----------
+        job : Job, mandatory
+            job
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+
+        Notes
+        -----------------
+        To send in the filename to the new save_results function in JobCadc
+        """
+        job.save_results(filename, verbose)
+
+    def login(self, user=None, password=None, certificate_file=None,
+              cookie_prefix=None, login_url=None, verbose=False):
+        """Performs a login.
+        User and password can be used or a file that contains user name and
+        password
+        (2 lines: one for user name and the following one for the password)
+        Parameters
+        ----------
+        user : str, mandatory if 'file' is not provided, default None
+            login name
+        password : str, mandatory if 'file' is not provided, default None
+            user password
+        certificate_file : str, mandatory if no 'user' & 'password' is provided
+            location of the certificate
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+
+        Notes
+        -----------------
+        Add certificates to the login
+        """
+        if certificate_file is None and user is None:
+            raise AttributeError(
+                'Input user/password or certificate file path')
+            return

```

Comment:
no returns after exceptions
```suggestion
```

Reply:
this is repeated throughout the function; I will not comment on further lines but they should also be changed